Add Create a backend add-on section in the Developer guide with `GenericSetup`

stevepiercy opened the pull request:
- Enhance the Development Guide index
- Update the term Buildout in the Glossary to point to dusty old, yet useful, docs in Plone 4.
- Source for GenericSetup already converted to MyST: https://github.com/plone/documentation/blob/5.2-myst/5.2-myst/develop/addons/components/genericsetup.md
- List of possible sources from https://github.com/plone/documentation/issues/1424

See https://github.com/plone/documentation/pull/1748#discussion_r1827660310

Closes https://github.com/plone/documentation/issues/1754.

Ping @petschki to take it from here.

<!-- readthedocs-preview plone6 start -->
----
📚 Documentation preview 📚: https://plone6--1757.org.readthedocs.build/

<!-- readthedocs-preview plone6 end -->

stevepiercy committed:
Add Create a backend add-on section in the Developer guide.
Ping @petschki
- Enhance the Development Guide index
- Update the term Buildout in the Glossary to point to dusty old, yet useful, docs in Plone 4.

(cherry picked from commit 8142b968a75133aca0bf0a958135b7604418cfb5)

stevepiercy committed:
- Update meta information
- Add term reference for Buildout

stevepiercy committed:
Merge branch '6.0' into dev-guide-create-a-backend-add-on

davisagli commented on docs/developer-guide/create-a-backend-add-on.md:
```diff
@@ -0,0 +1,21 @@
+# Create a backend add-on
+
+This section explains how a developer can create an {term}`add-on` for the Plone backend.
+You will use a framework called {term}`GenericSetup` and a tool called {term}`Buildout`.
```
I don't quite get your vision for this section yet.

Most of the documentation that can be written about how to create a backend add-on will be relevant regardless of how Plone was installed, and also regardless of whether the add-on was created from a Cookieplone template, bobtemplates.plone, or by hand.

So mentioning buildout here is off topic.

And learning about `GenericSetup` is a piece of what goes into creating an add-on, but it's certainly not the only or the most important piece.

stevepiercy replied:
I created this file to jump start something for @petschki to work on. I have no vision about this whatsoever. If the snippets aren't needed, they can be tossed out.

I moved the snippets here from another document that had them as headings in a how-to guide, instead of in a conceptual guide. IOW no one does "How to Buildout and GenericSetup".

stevepiercy committed:
Merge branch '6.0' into dev-guide-create-a-backend-add-on

stevepiercy committed:
Merge branch '6.0' into dev-guide-create-a-backend-add-on

stevepiercy committed:
Merge branch '6.0' into dev-guide-create-a-backend-add-on

stevepiercy committed:
Merge branch '6.0' into dev-guide-create-a-backend-add-on

stevepiercy committed:
Merge branch '6.0' into dev-guide-create-a-backend-add-on

stevepiercy committed:
Merge branch '6.0' into dev-guide-create-a-backend-add-on

stevepiercy commented:
I'm ready to close this issue as won't fix. Does anyone on @plone/classicui-team care about the topic of how to add or create a backend add-on section in the Developer guide with GenericSetup? If so, speak up, and let's get this finished.

petschki committed:
Merge branch '6.0' into dev-guide-create-a-backend-add-on

petschki commented:
@stevepiercy will have a look at this.

petschki committed:
create backend addon documentation

petschki committed:
create genericsetup documentation

petschki commented:
I've created a basic "addon creation documentation" and moved the old docs for GenericSetup here. I've also removed several outdated parts/link and tried to update the XML examples. Not finished yet though.

I would recommend to place these docs all inside "Backend" and make a link from "Developer guide / Develop backend add-ons" -> "Backend / Add-ons" (like its done for Volto add-ons)

Further we can add references for several `plonecli` installation guides (like Classic-UI / theming / create theme addon) to the new "Backend / Add-on" doc ...

What do you think?

petschki commented:
what's also missing is something like "Installing backend Add-on" ... at least I didn't find it... would also make sense to add it to "Backend / Add-ons" chapter.

EDIT: I mean adding a pypi package or source checkout to cookieplone/buildout project ...

petschki committed:
remove note

stevepiercy commented:
> what's also missing is something like "Installing backend Add-on" ... at least I didn't find it... would also make sense to add it to "Backend / Add-ons" chapter.
> 
> EDIT: I mean adding a pypi package or source checkout to cookieplone/buildout project ...

Was this what you couldn't find?

https://6.docs.plone.org/admin-guide/add-ons.html

The plan is to merge the Admin section into Develop section so that people don't have to think of what role they're in versus the task they want to perform. See https://github.com/plone/documentation/issues/1997.

stevepiercy commented on docs/i18n-l10n/translating-text-strings.md:
```diff
@@ -56,7 +56,7 @@ Information in the PO file headers is ignored.
 [`i18ndude`](https://pypi.org/project/i18ndude/) should be used to create a script which searches particular packages for translation strings.
 
-If you have created your add-on using [bobtemplates.plone](https://pypi.org/project/bobtemplates.plone/), then you will already have a script `update.sh` inside your package and a script `update_locale` in your buildout to extract the messages from your code.
+If you have created your add-on using [bobtemplates.plone](https://pypi.org/project/bobtemplates.plone/), then you will already have a script `update.sh` inside your package and a script `update_locale` in your {term}`buildout` to extract the messages from your code.
```
```suggestion
If you have created your add-on using [bobtemplates.plone](https://pypi.org/project/bobtemplates.plone/), then you will already have a script {file}`update.sh` inside your package and a script `update_locale` in your {term}`buildout` to extract the messages from your code.
```

stevepiercy committed:
Use file markup

petschki commented:
Ah ... exactly. 

Merging this to development section makes completely sense to me. Maybe we can do it like Volto does, to have an `Developer Guide -> Add-ons` directory, with `create/add/test" sub chapters.

Sidenote: what confuses me is, that when I click "Developer Guide -> Develop Volto Add-ons" I get redirected directly to the Volto Development 2nd level subdirectory ... from a UI perspective I would never redirect directly when clicking in the navigation area, instead have at least a small introductionary text in the content area with the link to the desired target.

stevepiercy commented:
> Sidenote: what confuses me is, that when I click "Developer Guide -> Develop Volto Add-ons" I get redirected directly to the Volto Development 2nd level subdirectory ... from a UI perspective I would never redirect directly when clicking in the navigation area, instead have at least a small introductionary text in the content area with the link to the desired target.

In a perfect world, we'd have one repo for all documentation in an easy-to-organize hierarchy. However, we use submodules to pull in documentation from multiple repos, and therefore can't easily have a single hierarchical structure. plone.restapi and plone.api have similar things.

We made the decision to use redirects as a compromise. There are placeholder files there that perform the actual redirect when the meta refresh tag fails in a browser for security reasons, but people were annoyed that they had to click twice to get to the content. Eventually, when the re-organization dust settles, we'll look into symlinks.

stevepiercy commented:
@petschki it's a holiday weekend here in the US. I'll review this next week.

Did you copy and paste this from the sources listed in the description, or did you modify any of it? It needs some serious editorial work.

petschki commented:
I've copied the merged GenericSetup document and removed the outdated parts ... I'll finish that this week. Happy Thanksgiving ...

jensens commented on docs/i18n-l10n/translating-text-strings.md:
```diff
@@ -273,7 +273,7 @@ This script hooks into the release process and builds the MO files for you.
 ### Installing i18ndude
 
-The recommended method is to have {term}`i18ndude` installed via your [buildout](https://www.buildout.org/en/latest/).
+The recommended method is to have {term}`i18ndude` installed via your buildout.
```
really?  I am not sure how far we yet adopted uv in the docs. But I would use it like so theses days:
```console
$ uvx i18ndude
usage: i18ndude [-h] {find-untranslated,rebuild-pot,merge,sync,filter,admix,list,trmerge} ...

i18ndude performs various tasks related to ZPT's, Python Scripts
and i18n.

Its main task is to extract translation strings (msgids) into a
.pot file (with the 'rebuild-pot' command), and sync the .pot file
with .po files (with the 'sync' command).

Call i18ndude with one of the listed subcommands followed by
--help to get help for that subcommand.

options:
  -h, --help            show this help message and exit

subcommands:
  {find-untranslated,rebuild-pot,merge,sync,filter,admix,list,trmerge}
```

stevepiercy replied:
@erral can you comment on this? If we're switching from buildout to uv, then the i18n documentation needs to be updated in a separate PR. This PR is focused on GenericSetup, and this change is only to remove the link to the shitty buildout docs in favor of those in Plone 4, believe it or not.

petschki replied:
@jensens I'd say yes of course, but this file has nothing to do with this PR ... @stevepiercy lets revert this change and keep focus on Add-ons here ...

erral replied:
When using a project scaffold created by cookieplone, one can run `make i18n` which does all the collecting of translatable strings.

erral commented on docs/developer-guide/create-a-backend-add-on.md:
````diff
@@ -0,0 +1,231 @@
+
+```shell
+uvx plonecli create addon <addon namespace package>
+```
````
I would remove using plonecli to create the addon. The latests alpha release calls `uvx cookieplone` under the hood, so it does the same.

We keep the `addon` template in there because all the testing is based on creating the addon

erral commented on docs/developer-guide/create-a-backend-add-on.md:
````diff
@@ -0,0 +1,231 @@
+ All templates below `addon` can be added to your newly created addon with:
+
+ ```shell
+ uvx ploncli add <subtemplate>
````
```suggestion
 uvx plonecli add <subtemplate>
```

erral commented on docs/developer-guide/create-a-backend-add-on.md:
````diff
@@ -0,0 +1,231 @@
+ ```shell
+ uvx ploncli add <subtemplate>
+ ```
+
````
You can also use `make add <subtemplate>` inside the `backend` folder of the structure created by cookieplone

petschki replied:
the Makefile only exists in the monorepo `backend` folder, am I correct? 

If you simply create a `uvx cookieplone backend_addon` there is no Makefile (at least for me) ... I'll outline both scenarios.

erral replied:
I have just made a test, I see the correct `Makefile` and it contains the `make add` command support.

petschki replied:
Oh, sorry... you're right. I've looked in my `plonecli` created addon, not in the `cookieplone backend_addon` version 🫣

I'll update this ...

stevepiercy committed:
Revert changes in i18n and glossary to keep PR focused on GenericSetup

stevepiercy committed:
Revert changes in i18n and glossary to keep PR focused on GenericSetup

stevepiercy committed:
Merge branch '6.0' into dev-guide-create-a-backend-add-on

petschki committed:
Update docs/developer-guide/create-a-backend-add-on.md

Co-authored-by: Mikel Larreategi <mlarreategi@codesyntax.com>

petschki committed:
remove plonecli addon creation and note for monorepo "make" command

petschki committed:
only mention "make" command for "cookieplone" generated addons

erral commented on docs/developer-guide/create-a-backend-add-on.md:
```diff
@@ -0,0 +1,157 @@
+
+## Add `plonecli` subtemplate to an addon
+
+The generated addon contains a {file}`bobtemplates.cfg` file which lets you add several subtemplates with `plonecli`.
```
```suggestion
The generated addon contains a {file}`bobtemplate.cfg` file which lets you add several subtemplates with `plonecli`.
```

stevepiercy commented:
@petschki @erral are there any other technical issues to resolve before I dive in for a general editorial review? Please let me know. Thank you!

petschki committed:
Update docs/developer-guide/create-a-backend-add-on.md

Co-authored-by: Mikel Larreategi <mlarreategi@codesyntax.com>

petschki commented:
> @petschki @erral are there any other technical issues to resolve before I dive in for a general editorial review? Please let me know. Thank you!

I still need to go through the genericsetup doc technically. There are some outdated bits which have to be removed.

MrTango committed:
Merge branch '6.0' into dev-guide-create-a-backend-add-on

stevepiercy commented:
> I still need to go through the genericsetup doc technically. There are some outdated bits which have to be removed.

@petschki I don't see a commit to remove the outdated bits. Did you forget to push it? Please let me know. Thank you!

MrTango commented:
I would use plonecli for all, after @erral did the final release in the comming days.
It will then work as it was working for Plone 5 development, with the difference that now the create addon part is using cookieplone backend_addon under the hood. This is exactly why we have plonecli, to hide stuff under it and give the developers a clean cli.

example:

```sh
uv tool install plonecli
plonecli create addon collective.myaddon

cd collective.myaddon
plonecli add content_type
...
```

`make add`, would be an abstraction, which calls an abstraction (plonecli), which calls a bobtemplate.
I don't mind that the make command exist, for who want to do it this way, but i would not put this out as the default, otherwise we can take out plonecli from the mix and do it purely in Make.

We could even create a full cookieplone project with plonecli.

```sh
plonecli create volto-project
```

or

```sh
plonecli create cookieplone-project
```

or

```sh
plonecli create plone-project
```

it doesn't matter and we are flexible to replace part under it, with the next cool tool, someone creates, without destroing exiting docs and api's.

my 50cent...

petschki commented:
@MrTango I had documente the `plonecli` part initially and removed it as @erral's request ... can we find a consent here?

I think, due to the fact that the whole plone installation and theme development documentation uses cookieplone, we should at least use it for the inital addon package creation too ... when adding further features inside the addon, we can document this with using `plonecli` ...

erral commented:
@petschki @MrTango sorry for the mess...

with plonecli right now we can create an add-on calling `plonecli create addon` which creates a `backend_addon` calling `uvx cookieplone backend_addon`.

I haven't checked to do the same for volto addon, project, etc, but could be done, and it is not that hard.

If we do that we can document everything to use plonecli indeed, and `cookieplone` would be just an "implementation detail".

But we also have documentation to create Volto projects using `uvx cookieplone` so the mess is bigger...

yurj commented:
cookieplone is the framework with utils to full manage a Plone project while plonecli is a client that implements the frameworks (past and present).

stevepiercy commented:
What are the use cases that we're trying to solve with this page of documentation? Maybe that answer will help us give a direction. Would y'all agree there are essentially two scenarios?

1. Create a backend add-on only.
2. Create a backend add-on along with a combination of anything else, such as a full Plone project, a Volto add-on, project documentation scaffold, and devops.

Neither Cookieplone nor plonecli are going away anytime soon. If we have two or more ways of doing the same thing, that's fine, too, provided we give sufficient information for the reader to choose which tool is best for them. For an example of that, see [Choose a user interface](https://6.docs.plone.org/conceptual-guides/choose-user-interface.html). It's also OK if we use one tool to do one part, such as using Cookieplone to create a monorepo with all the things, and another tool to do another part, such as add new features to the backend add-on.

Does this help at all?

petschki committed:
explain how to create a backend-addon with plonecli only.

petschki commented:
@erral @MrTango I've now replaced the `cookieplone` part with my formerly removed `plonecli` part ...

petschki committed:
minor command fix

stevepiercy committed:
Merge branch '6.0' into dev-guide-create-a-backend-add-on

stevepiercy committed:
Begin editorial clean up

stevepiercy commented on docs/backend/generic-setup.md:
````diff
@@ -0,0 +1,1755 @@
+```{todo}
+remove archetypes example code everywhere
+```
+
````
Either remove this todo or remove the code to which it refers.

davisagli commented on docs/backend/generic-setup.md:
I just skimmed through to check that this is still accurate/relevant information. It is and we should prioritize getting this merged. It does need some editing to match current doc styles.

davisagli commented on docs/developer-guide/create-a-backend-add-on.md:
````diff
@@ -0,0 +1,149 @@
+```
+
+```console
+> uvx plonecli create addon collective.addon
````
I tried this and here's what happened:
```
$ uvx plonecli create addon collective.addon
Traceback (most recent call last):
  File "/Users/davisagli/.local/share/uv/tools/plonecli/bin/plonecli", line 4, in <module>
    from plonecli.cli import cli
  File "/Users/davisagli/.local/share/uv/tools/plonecli/lib/python3.14/site-packages/plonecli/__init__.py", line 4, in <module>
    import pkg_resources
ModuleNotFoundError: No module named 'pkg_resources'
```

- uv picked the latest Python it knows about, which is 3.14. Does plonecli support it?
- it got the new version of setuptools which doesn't have pkg_resources

It looks like @erral already fixed the latter in https://github.com/plone/plonecli/commit/a00e960e9507bd2a7e1caa835a9bef3d0781fad0 ... could that be released?

erral replied:
Yes sorry, I didn't find time to request releases of all of them (mrbob, plonecli, bobtemplates.plone). I'll do it later today.

erral replied:
This is already working.

davisagli reviewed (commented):
Let's see if we can get this over the finish line.

petschki committed:
Merge branch '6.0' into dev-guide-create-a-backend-add-on

erral commented:
I have just released new versions for mr.bob, plonecli and bobtemplates removing all pkg_resources style namespaces. I tested the three of them together and it works as expected. I will review this PR tomorrow to check that everything is Ok.

erral committed:
Merge branch '6.0' into dev-guide-create-a-backend-add-on

erral commented on docs/developer-guide/create-a-backend-add-on.md:
````diff
@@ -0,0 +1,149 @@
+
+
+isort-apply: successful:
+isort-apply: install_deps> python -I -m pip install isort -c constraints.txt
+isort-apply: commands[0]> isort /Users/<username>/Development/collective.addon/src
+/Users/<username>/Development/collective.addon/setup.py
+Fixing /Users/<username>/Development/collective.addon/src/collective/addon/testing.py
+Fixing /Users/<username>/Development/collective.addon/src/collective/addon/tests/test_setup.py
+  isort-apply: OK (2.57=setup[1.94]+cmd[0.63] seconds)
+  congratulations :) (2.59 seconds)
+
+
+Identified `/` as project root containing a file system root.
+Sources to be formatted: "Users/<username>/Development/collective.addon/src",
+  "Users/<username>/Development/collective.addon/setup.py"
+src/collective/__init__.py wasn't modified on disk since last run.
+src/collective/addon/browser/__init__.py wasn't modified on disk since last run.
+src/collective/addon/locales/__init__.py wasn't modified on disk since last run.
+src/collective/addon/tests/__init__.py wasn't modified on disk since last run.
+src/collective/addon/interfaces.py already well formatted, good job.
+reformatted src/collective/addon/__init__.py
+reformatted src/collective/addon/setuphandlers.py
+reformatted src/collective/addon/testing.py
+reformatted setup.py
+reformatted src/collective/addon/locales/update.py
+reformatted src/collective/addon/tests/test_setup.py
+
+All done! ✨ 🍰 ✨
+6 files reformatted, 5 files left unchanged.
+
+black-enforce: successful:
+black-enforce: install_deps> python -I -m pip install black -c constraints.txt
+black-enforce: commands[0]> black -v src setup.py
+  black-enforce: OK (2.60=setup[2.12]+cmd[0.48] seconds)
+  congratulations :) (2.61 seconds)
+
+
+git init is disabled!
+Generated file structure at /Users/<username>/Development/collective.addon/collective.addon
+```
````
```suggestion
RUN: bobtemplates.plone:addon -O collective.addon

Welcome to mr.bob interactive mode. Before we generate directory structure, some questions need to be answered.

Answer with a question mark to display help.
Values in square brackets at the end of the questions show the default value if there is no answer.


--> Author's name [Mikel Larreategi]: 

--> Author's email [mlarreategi@codesyntax.com]: 

--> Author's GitHub username: 

--> Package description [An add-on for Plone]: 

--> Do you want me to initialize a GIT repository in your new package? (y/n) [y]: 

--> Plone version [6.0.0]: 

--> Python version for virtualenv [python3]: 

--> Do you want me to activate VS Code support? (y/n) [y]: 


╭──────────────────────────────── cookieplone ─────────────────────────────────╮
│                                                                              │
│                                   *******                                    │
│                               ***************                                │
│                             ***             ***                              │
│                           ***    ***          ***                            │
│                          ***    *****          ***                           │
│                         ***      ***            ***                          │
│                         ***               ***   ***                          │
│                         ***              *****  ***                          │
│                         ***      ***      ***   ***                          │
│                          ***    *****          ***                           │
│                           ***    ***          ***                            │
│                             ***             ***                              │
│                               ***************                                │
│                                   *******                                    │
│                                                                              │
╰──────────────────────────────────────────────────────────────────────────────╯
╭──────────────────────────────── Plone Addon ─────────────────────────────────╮
│                                                                              │
│ Creating a new Plone Addon                                                   │
│                                                                              │
│ Sanity check results:                                                        │
│                                                                              │
│   - Cookieplone: ✓                                                           │
│   - uv: ✓                                                                    │
│   - git: ✓                                                                   │
│                                                                              │
╰──────────────────────────────────────────────────────────────────────────────╯
 -> Remove files used in classic UI setup
 -> Create namespace packages
 -> Format code
 -> Initialize Git repository
 -> Generate documentation scaffold
 -> Setup VSCode configuration
 -> Setup GitHub CI
╭────────────────────────── New addon was generated ───────────────────────────╮
│                                                                              │
│ collective.addon                                                             │
│                                                                              │
│ Now, enter the repository, start coding, and push to your organization.      │
│                                                                              │
│ Sorry for the convenience,                                                   │
│ The Plone Community.                                                         │
│                                                                              │
│ https://plone.org/                                                           │
╰──────────────────────────────────────────────────────────────────────────────╯
Done
Generated file structure at /tmp/collective.addon
```

stevepiercy replied:
Thanks @erral! The terminal output is very wide, though. The code block area when rendered in the docs starts to wrap at the 90th character, requiring users to horizontal scroll back and forth to view the output. Would you please narrow your terminal session window and run it again?

erral replied:
OK, let's try again

erral reviewed (changes requested):
I have checked the output of `uvx plonecli` and it is a bit different than the previous version.

stevepiercy reviewed (commented):
I started a review, and I'm not sure what remains in here about archetypes. Anyone know?

I'm also slowly going through this to pass editorial review, which will be in subsequent commits.

stevepiercy commented on docs/backend/generic-setup.md:
```diff
@@ -0,0 +1,1750 @@
+  Example:
+  You have branch master with version 1025.
+  You make backwards incompatible changes and you increase the version to 2000.
+  You create a maintenance branch where the next metadata version will be 1026.
```
This section is too much. I think it should be rewritten to promote Python version specifiers as described in https://packaging.python.org/en/latest/specifications/version-specifiers/, and mention that if there are integer numbers or none, then use Python version specifiers.

Would that break anything? @mauritsvanrees or anyone else with historical knowledge on this topic.

davisagli replied:
@stevepiercy I think this section is an accurate, if a bit informal, description of how GenericSetup profile versions are used in practice. I have never seen one that is not an integer....this claims that works, but I can't confirm that without trying it. I would be very surprised if more complex Python version specifiers (with e.g. dev/a/b/rc) worked here.

stevepiercy replied:
@davisagli thanks for the feedback. This is what I latched onto.

> ... Versions with dots work fine too.
> 
> They get ordered just like they would when used for packages on PyPI.

Which implies the sort order is by Python version specifiers, not alphabetic, numeric, or alphanumeric. But then a version with a dot could also be a decimal. Agh! GenericSetup's change log was no help to me either.

It would be nice to be certain of what is actually supported. If certainty is not possible, then I can simplify this section to specify integers only, while dropping mention of dots or PyPI versions for metadata versions. It might not be complete, but accuracy and clarity take precedence.

davisagli replied:
I checked the code. It does compare versions normalized using `packaging.version.parse`. So it's right to say we accept Python version specifiers here. Which means your edits in https://github.com/plone/documentation/pull/2083 are mostly good already. I'll update it to not talk about integers as a legacy edge case, because they are still the most commonly used subset of Python version specifiers in this context.

stevepiercy commented on docs/backend/generic-setup.md:
````diff
@@ -0,0 +1,1750 @@
+      />
+
+</configure>
+```
````
This section is also confusing. First we say "do this", then later say "do this best practice." This section should be rewritten to provide a single example as _the best way_ to do it _first_, then follow it with variations in subsequent examples.

stevepiercy committed:
Rewrite and restructure the GenericSetup documentation to have a proper introduction and overview section.
Add Zope Management Interface as an alias to ZMI in the glossary and use it as a term.

stevepiercy committed:
Tidy up links

stevepiercy committed:
Clean up Create a profile section

stevepiercy committed:
Clean up Multiple profiles section, removing obsolete mention of 5.0 and older

stevepiercy committed:
Clean up "Add-on specific issues"
- Rename to "Add-on properties", as these aren't issues at all.
- Fix incorrect cross-reference link to custom installation steps
- Rename `Import various` step to the step actually used, `setup_various`
- Create a seealso directive to a really old web archive. Maybe remove this link?

stevepiercy committed:
Clean up "List available profiles"
- Move comment from code block to narrative text.
- Use python lexer in code block.

stevepiercy committed:
Clean up "Install a profile" and its subsections
- Clarify purpose of this section with definitive language instead of vague language.
- `plone.app.testing` fix link to current documentation on GitHub.
- Rename "Manually" section to explicit "Manual installation"
- Use python lexer in code block.
- Change `note` to `versionchanged` directive and clarify wording.

stevepiercy committed:
Add a `versionchanged` directive and clarify wording to "Multiple profiles" section.

stevepiercy committed:
Grammar, MyST syntax, and style corrections

stevepiercy committed:
Revise "Metadata version numbers" to modernize it to a best practice placed first, and remove unnecessary and obsolete information.

stevepiercy committed:
Revise "Metadata version numbers" to modernize it to a best practice placed first, and remove unnecessary and obsolete information.

stevepiercy committed:
Revise "Custom installer code"
- Improve grammar, syntax, and style
- It's still way too confusing and needs technical help.

stevepiercy committed:
- Clean up code comments
- Fix reference target syntax
- Replace Management Interface with ZMI
- Add a TODO to fix the autodoc warning.

stevepiercy committed:
- Fix typos

stevepiercy reviewed (commented):
After performing an editorial review and creating a separate PR against this one, then reading through these latest two sections, there is no way that this PR can be merged. It refers to obsolete practices, things that don't actually work, and it wanders without clear direction or context.

I understand it's the result of decades of copy paste and ad hoc contributions. It's going to take a lot of collaboration between those with the technical and historical knowledge and me to make it understandable for readers who don't have decades of experience with Plone.

To move this forward, please take a look at https://github.com/plone/documentation/pull/2083/ and add your comments there. Thank you!

stevepiercy committed:
Fix reference target

stevepiercy committed:
Fix invalid XML syntax

stevepiercy committed:
Fix invalid reference links

stevepiercy committed:
Fix broken automodule member list. However, now it reveals malformed docstrings in the Products.GenericSetup.rolemap module.

```
/documentation/venv/lib/python3.13/site-packages/Products/GenericSetup/rolemap.py:docstring of Products.GenericSetup.rolemap.importRolemap:9: ERROR: Unexpected indentation. [docutils]
/documentation/venv/lib/python3.13/site-packages/Products/GenericSetup/rolemap.py:docstring of Products.GenericSetup.rolemap.importRolemap:19: WARNING: Definition list ends without a blank line; unexpected unindent. [docutils]
```

stevepiercy committed:
Fix broken automodule member list. However, now it reveals malformed docstrings in the Products.GenericSetup.rolemap module.

```
/documentation/venv/lib/python3.13/site-packages/Products/GenericSetup/rolemap.py:docstring of Products.GenericSetup.rolemap.importRolemap:9: ERROR: Unexpected indentation. [docutils]
/documentation/venv/lib/python3.13/site-packages/Products/GenericSetup/rolemap.py:docstring of Products.GenericSetup.rolemap.importRolemap:19: WARNING: Definition list ends without a blank line; unexpected unindent. [docutils]
```

stevepiercy committed:
Apply suggestions from code review

Co-authored-by: David Glick <david@glicksoftware.com>

davisagli committed:
Remove references to archetypes

davisagli committed:
Edits after my review

davisagli committed:
Remove rolemap automodule which is a distraction

davisagli committed:
fix ref

davisagli committed:
Merge pull request #2083 from plone/dev-guide-create-a-backend-add-on-editorial

Editorial review of Add Create a backend add-on section in the Developer guide with GenericSetup #1757

stevepiercy committed:
Remove old import various step, per https://github.com/plone/documentation/pull/2083#discussion_r3280243322

stevepiercy committed:
Add import per https://github.com/plone/documentation/pull/2083#discussion_r3280371012

stevepiercy committed:
Fix ref link

stevepiercy commented:
@davisagli would you please check these two commits that appear to have been omitted, per my commit messages?

- https://github.com/plone/documentation/pull/1757/commits/91497a054b5a7b96d6d0185892b0ed35d67e62ae
- https://github.com/plone/documentation/pull/1757/commits/025e91cd8b78438997f3f96215fdc1e93365a88e

...and let me know whether to revise them.

I'm going to start another branch off this one for rewriting the next chunk. Thank you for your help!

stevepiercy committed:
Merge branch '6.0' into dev-guide-create-a-backend-add-on

davisagli commented:
@stevepiercy Yeah those commits are good -- I remember making those changes, so I must have messed something up in the merging. I'm ready to do more review once you have another chunk ready.
